ts/components/conversation/media-gallery/groupMessagesByDate.ts

```diff
@@ -0,0 +1,164 @@
+/**
+ * @prettier
+ */
+import moment from 'moment';
+import { compact, groupBy, sortBy } from 'lodash';
+
+import { Message } from './propTypes/Message';
+// import { missingCaseError } from '../../../util/missingCaseError';
+
+type StaticSectionType = 'today' | 'yesterday' | 'thisWeek' | 'thisMonth';
+type YearMonthSectionType = 'yearMonth';
+
+interface GenericSection<T> {
+  type: T;
+  messages: Array<Message>;
+}
+type StaticSection = GenericSection<StaticSectionType>;
+type YearMonthSection = GenericSection<YearMonthSectionType> & {
+  year: number;
+  month: number;
+};
+export type Section = StaticSection | YearMonthSection;
+
+export const groupMessagesByDate = (
+  timestamp: number,
+  messages: Array<Message>
+): Array<Section> => {
+  const referenceDateTime = moment.utc(timestamp);
+  const today = moment(referenceDateTime).startOf('day');
+  const yesterday = moment(referenceDateTime)
+    .subtract(1, 'day')
+    .startOf('day');
+  const thisWeek = moment(referenceDateTime).startOf('isoWeek');
+  const thisMonth = moment(referenceDateTime).startOf('month');
+
+  const sortedMessages = sortBy(messages, message => -message.received_at);
+  const messagesWithSection = sortedMessages.map(
+    withSection({
+      today,
+      yesterday,
+      thisWeek,
+      thisMonth,
+    })
+  );
+  const groupedMessages = groupBy(messagesWithSection, 'type');
+  const yearMonthMessages = Object.values(
+    groupBy(groupedMessages.yearMonth, 'order')
+  ).reverse();
+  return compact([
+    toSection(groupedMessages.today),
+    toSection(groupedMessages.yesterday),
+    toSection(groupedMessages.thisWeek),
+    toSection(groupedMessages.thisMonth),
+    ...yearMonthMessages.map(group => toSection(group)),
+  ]);
+};
+
+const toSection = (
+  messagesWithSection: Array<MessageWithSection> | undefined
+): Section | null => {
+  if (!messagesWithSection || messagesWithSection.length === 0) {
+    return null;
+  }
+
+  const firstMessageWithSection: MessageWithSection = messagesWithSection[0];
+  if (!firstMessageWithSection) {
+    return null;
+  }
+
+  const messages = messagesWithSection.map(
+    messageWithSection => messageWithSection.message
+  );
+  switch (firstMessageWithSection.type) {
+    case 'today':
+    case 'yesterday':
+    case 'thisWeek':
+    case 'thisMonth':
+      return {
+        type: firstMessageWithSection.type,
+        messages,
+      };
+    case 'yearMonth':
+      return {
+        type: firstMessageWithSection.type,
+        year: firstMessageWithSection.year,
+        month: firstMessageWithSection.month,
```

Comment:
I just did a quick scan, and I didn't see where `.year` and `.month` were used anywhere else. Are they used?

Reply:
Yep, they’ll be used for rendering the year and month in each section header.